plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Currency.hs

```diff
@@ -0,0 +1,171 @@
+    Currency
+        { curRefTransactionOutput = (V.plcTxHash h, i)
+        , curRefAmount            = amt
+        , curRefTokenName         = mkTokenName n
+        }
+
+curValidator :: Currency -> ValidatorScript
+curValidator cur =
+    ValidatorScript (Ledger.applyScript mkValidator (Ledger.lifted cur)) where
+        mkValidator = Ledger.fromCompiledCode ($$(P.compile [||
+            let validate :: Currency -> () -> () -> V.PendingTx -> ()
+                validate (Currency (refHash, refIdx) amt token) () () p = 
+                    let 
+                        ownSymbol = $$(V.ownCurrencySymbol) p
+                        forged = $$(V.valueForged) p
+                        expected = $$(Value.singleton) ownSymbol token amt 
+
+                        -- True if the pending transaction forges the amount of
+                        -- currency that we expect
+                        forgeOK =
+                            let v = $$(Value.eq) expected forged
+                            in $$(P.traceIfFalseH) "Value forged different from expected" v
+
+                        -- True if the pending transaction spends the output
+                        -- identified by @(refHash, refIdx)@
+                        txOutputSpent = 
+                            let v = $$(V.spendsOutput) p refHash refIdx
+                            in  $$(P.traceIfFalseH) "Pending transaction does not spend the designated transaction output" v
+
+                    in
+                        if $$(P.and) forgeOK txOutputSpent
+                        then ()
+                        else $$(P.error) ($$(P.traceH) "Invalid forge" ())
+            in
+                validate
+            ||]))
+
+-- | Currency data script (unit value).
+curDataScript :: DataScript
+curDataScript = DataScript $ Ledger.lifted ()
+
+-- | Currency redeemer (unit value).
+curRedeemer :: RedeemerScript
+curRedeemer = RedeemerScript $ Ledger.lifted ()
+
+-- | The address of a 'Currency' contract.
+curAddress :: Currency -> Address
+curAddress = Ledger.scriptAddress . curValidator
+
+-- | The 'Value' forged by the 'curValidator' contract
+forgedValue :: Currency -> Value
+forgedValue cur = 
+    let a = plcCurrencySymbol (curAddress cur)
+        t = curRefTokenName cur
+        i = curRefAmount cur
+    in
+        $$(Value.singleton) a t i
+
+-- | @forge c n@ forges @n@ units of a currency called @c@ and
+--   pays them to a public key address owned by the wallet.
+forge :: (WalletAPI m, WalletDiagnostics m) => String -> Int -> m Currency
+forge nm amount = do
+    pk <- WAPI.ownPubKey
+
+    -- 1. We need to create the reference transaction output using the 
+    --    'PublicKey' contract. That way we get an output that behaves
+    --    like a normal public key output, but is not selected by the
+    --    wallet during coin selection. This ensures that the output still 
+    --    exists when we spend it in our forging transaction.

```

Comment:
:thinking: Does coin selection *remove* public key outputs? That seems bad...

Reply:
Wouldn't it be fine if it just left them alone and only added outputs to balance things?

Reply:
No it doesn't remove them, that's the problem. If we pick a public key output before submitting the first transaction, then the wallet may select that same public key output to pay for the first transaction so it won't be available for the second transaction. We can't earmark pubkey outputs for future use.

Reply:
Ah I see. So coin selection treats all pubkey outputs as available, so we have to hide it in a special transaction.